### PR TITLE
bugfix: Deps hangs when using relative paths via `--project-dir`

### LINKED ISSUE #7491: [Bug] dbt deps --project-dir $DBT_PROJECT_DIR hangs with dbt=1.5.0
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

After upgrading dbt-core to 1.5.0, our CI jobs hangs after running the following command from the parent directory of `integration_tests`:
```
dbt deps --project-dir integration_tests
```
Output:
```
#!/bin/bash -eo pipefail
. venv/bin/activate
  dbt deps --project-dir $DBT_PROJECT_DIR
01:29:08  Running with dbt=1.5.0
```
This will eventually result in a timeout [error](https://app.circleci.com/pipelines/github/calogica/dbt-expectations/253/workflows/92c6f533-2054-4b39-b033-ca8d77701187/jobs/177/parallel-runs/0/steps/0-103).

This was previously working under dbt version 1.4.1 and is currently blocking CI jobs for `dbt-expectations` and `dbt-date`.

If you `cd integrations_tests` and then run `dbt deps`, this command succeeds:

```
integration_tests git:(main) dbt deps
01:37:38  Running with dbt=1.5.0
01:37:39  Installing ../
01:37:39  Installed from <local @ ../>
01:37:39  Installing calogica/dbt_date
01:37:39  Installed from version 0.7.2
01:37:39  Up to date!
``` 

### Expected Behavior

This was previously working under dbt version 1.4.1 and should simply install the packages defined in `integration_tests/packages.yml`.

### Steps To Reproduce

Using dbt=1.5.0, in a project that contains a subproject, such as dbt-utils or dbt-expectations, run the following from the root folder:

`dbt deps --project-dir subproject` where `subproject` is the name of the folder containing the subproject.

### Relevant log output

_No response_

### Environment

```markdown
- OS: MacOS 13.3.1
- Python: 3.9
- dbt: 1.5.0
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by iknox-fa**
resolves #7491
### Description
When a relative path is provided by using `--project-dir` the resolution of that path would fail in deps because we change the working directory during the task's lifecycle.  This means that we got invalid paths when trying to fetch the metadata for `LocalPinnedPackage`s because it uses `system.resolve_path_from_base` which in turn relies on `os.path.abspath` to properly resolve the full path, which it can't do correctly because said path is now resolved using a different `cwd` than when originally defined.

If you can follow that convoluted logic give yourself a pat on the back.  :)

In the long run we shouldn't do things like `os.cwd()` in our tasks, but it's pretty deeply built into how we handle the "portable" nature of things like manifests and whatnot so it's a ticket for another day.  

See #7615 for tracking larger issues around file paths.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by stu-k — APPROVED** [review]
This seems like a reasonable short-term implementation and you linked the spike to look into how we use paths more broadly 👍

File: tests/functional/dependencies/test_local_dependency.py [hunk 1]
```diff
@@ -155,23 +152,16 @@ def test_local_dependency(self, project):
...
-        # prove model can run when importing that dependency
-        local_path = Path("models") / "my_model.sql"
-        results = run_dbt(
-            ["run", "--models", f"+{local_path}"],
-        )
-        assert len(results) == 2
+class TestSimpleDependencyRelativePath(BaseDependencyTest):
+    def test_local_dependency_relative_path(self, project):
+        last_dir = Path(project.project_root).name
+        os.chdir("../")
```
[the remark this answers is not in the document]
Comment (iknox-fa) [reply tests/functional/dependencies/test_local_dependency.py:159]: I could, but we don't do that in the codebase so this is a more accurate test without it.

File: core/dbt/task/deps.py [hunk 2]
```diff
@@ -29,6 +29,12 @@
 
 class DepsTask(BaseTask):
     def __init__(self, args: Any, project: Project):
+        # N.B. This is a temporary fix for a bug when using relative paths via
+        # --project-dir with deps.  A larger overhaul of our path handling methods
+        # is needed to fix this the "right" way.
+        # See GH-7615
+        project.project_root = str(Path(project.project_root).resolve())
```
[the remark this answers is not in the document]
Comment (iknox-fa) [reply core/dbt/task/deps.py:36]: That wouldn't work.  `is_dir` [returns a bool](https://docs.python.org/3/library/pathlib.html#pathlib.Path.is_dir), whereas `resolve()` [returns a Path](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve)

File: tests/functional/dependencies/test_local_dependency.py [hunk 3]
```diff
@@ -155,23 +152,16 @@ def test_local_dependency(self, project):
             [f"{project.test_schema}.dep_source_model", f"{project.test_schema}.seed"],
         )
 
-    def test_no_dependency_paths(self, project):
```
[the remark this answers is not in the document]
Comment (iknox-fa) [reply tests/functional/dependencies/test_local_dependency.py:158]: Whoops!  Good catch.

File: tests/functional/dependencies/test_local_dependency.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (iknox-fa) [reply tests/functional/dependencies/test_local_dependency.py:159]: On second though the chdir here doesn't effect the internal one.  Sure, we'll wrap it.

File: tests/functional/dependencies/test_local_dependency.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbeatty10) [reply tests/functional/dependencies/test_local_dependency.py:158]: This test is added back in https://github.com/dbt-labs/dbt-core/pull/7642 and https://github.com/dbt-labs/dbt-core/pull/7649

**Pull request merged**